certora/checker/create_certificate.py

```diff
@@ -0,0 +1,82 @@
+import sys
+import json
+from web3 import Web3, EthereumTesterProvider
+
+w3 = Web3(EthereumTesterProvider())
+
+
+# Returns the hash of a node given the hashes of its children.
+def keccak_node(left_hash, right_hash):
+    return w3.to_hex(
+        w3.solidity_keccak(["bytes32", "bytes32"], [left_hash, right_hash])
+    )
+
+
+# Returns the hash of a leaf given the rewards details.
+def keccak_leaf(address, amount):
+    address = w3.to_checksum_address(address)
+    return w3.to_hex(w3.solidity_keccak(["address", "uint256"], [address, amount]))
+
+
+certificate = {}
+hash_to_address = {}
+hash_to_value = {}
+left = {}
+right = {}
+
+
+# Populates the fields of the tree along the path given by the proof.
+def populate(address, amount, proof):
+    amount = int(amount)
+    computedHash = keccak_leaf(address, amount)
+    hash_to_address[computedHash] = address
+    hash_to_value[computedHash] = amount
+    for proofElement in proof:
+        [leftHash, rightHash] = (
+            [computedHash, proofElement]
+            if computedHash <= proofElement
+            else [proofElement, computedHash]
+        )
+        computedHash = keccak_node(leftHash, rightHash)
+        left[computedHash] = leftHash
+        right[computedHash] = rightHash
+        hash_to_address[computedHash] = keccak_node(computedHash, computedHash)[:42]
+
+
+# Traverse the tree and generate corresponding instruction for each internal node and each leaf.
+def walk(node):
+    if node in left:
```

Comment:
I'm still not sure to understand this. the root is a key of left, but not a value of left right ?

left[computedHash] = leftHash

Reply:
This function goes up, following the given proof, and keeping track of the current hash `computedHash`. At each step, you have a `leftHash` and a `rightHash`, where those are taken from the current hash and the proof element, and which one is which depends on whether you come from a right child or a left child. Those allow you to compute the parent hash, that now takes the place of `computedHash`. At this point we do
```
left[computedHash] = leftHash
```
because the (hash of the) left child of `computedHash` is `leftHash`.
When we arrive at the very top, `computedHash` is the root, `leftHash` is its left child, and `rightHash` is its right child, and we do the same thing: `left[computedHash] = leftHash`.